qbraid_qir/qasm3/visitor.py

```diff
@@ -0,0 +1,828 @@
+
+        # so we need to update the arguments of these gate applications with the actual
+        # qubit identifiers and then RECURSIVELY call the visit_generic_gate_operation
+        qubit_map = {
+            formal_arg.name: actual_arg
+            for formal_arg, actual_arg in zip(gate_definition.qubits, op_qubits)
+        }
+        param_map = {
+            formal_arg.name: actual_arg
+            for formal_arg, actual_arg in zip(gate_definition.arguments, operation.arguments)
+        }
+        for gate_op in gate_definition.body:
+            # necessary to avoid modifying the original gate definition
+            # in case the gate is reapplied
+            gate_op_copy = copy.deepcopy(gate_op)
+            if isinstance(gate_op, QuantumGate):
+                self._transform_gate_params(gate_op_copy, param_map)
+                self._transform_gate_qubits(gate_op_copy, qubit_map)
+                self._visit_generic_gate_operation(gate_op_copy)
+            elif isinstance(gate_op, QuantumMeasurementStatement):
+                self._print_err_location(gate_op.span)
+                raise Qasm3ConversionError("Unsupported measurement statement in gate definition")
+            elif isinstance(gate_op, QuantumReset):
+                self._print_err_location(gate_op.span)
+                raise Qasm3ConversionError("Unsupported reset statement in gate definition")
+            else:
+                # TODO: add control flow support
+                self._print_err_location(gate_op.span)
+                raise Qasm3ConversionError(f"Unsupported gate definition statement {gate_op}")
+
+    def _visit_generic_gate_operation(self, operation: QuantumGate) -> None:
+        """Visit a gate operation element.
+
+        Args:
+            operation (QuantumGate): The gate operation to visit.
+
+        Returns:
+            None
+        """
+        if operation.name.name in self._custom_gates:
+            self._visit_custom_gate_operation(operation)
+        else:
+            self._visit_basic_gate_operation(operation)
+
+    def _visit_classical_operation(self, statement: ClassicalDeclaration) -> None:
+        """Visit a classical operation element.
+
+        Args:
+            statement (ClassicalType): The classical operation to visit.
+
+        Returns:
+            None
+        """
+        decl_type = statement.type
+
+        if isinstance(decl_type, (qasm3IntType, UintType, FloatType)):
+            size = decl_type.size if decl_type.size is not None else 1
+
+            # we only support static integer sizes
+            if isinstance(size, IntegerLiteral):
+                value = size.value
+            else:
+                raise Qasm3ConversionError(f"Unsupported integer size {size} in {statement}")
+            var_name = statement.identifier.name
+
+            # how to add this in the QIR???
+        else:
+            raise Qasm3ConversionError(f"Unsupported classical type {decl_type} in {statement}")
```

Comment:
What's going on in this function? If not implemented, we can just raise `NotImplementedError`. Either way, will want to clean this up a bit before we merge.

Reply:
I was exploring how to define classical variables like `int`, `bool`, etc. but have not implemented it yet. Will just raise `NotImplemented` error as of now